common/src/balance_unit.rs

```diff
@@ -0,0 +1,570 @@
+            is_divisible,
+        }
+    }
+
+    pub fn divisible(balance: Balance) -> Self {
+        Self::new(balance, true)
+    }
+
+    pub fn indivisible(balance: Balance) -> Self {
+        Self::new(balance, false)
+    }
+
+    pub fn copy_divisibility(&self, balance: Balance) -> Self {
+        Self::new(balance, self.is_divisible)
+    }
+
+    pub fn get(&self) -> &Balance {
+        &self.inner
+    }
+
+    pub fn set(&mut self, value: Balance) {
+        self.inner = value
+    }
+
+    pub fn is_divisible(&self) -> bool {
+        self.is_divisible
+    }
+
+    pub fn pow(&self, x: u32) -> Result<Self, ArithmeticError> {
+        let balance = if self.is_divisible {
+            FixedWrapper::from(self.inner).pow(x).try_into_balance()?
+        } else {
+            self.inner.checked_pow(x).ok_or(ArithmeticError::Overflow)?
+        };
+        Ok(Self::new(balance, self.is_divisible))
+    }
+
+    pub fn sqrt(&self) -> Result<Self, ArithmeticError> {
+        let balance = if self.is_divisible {
+            FixedWrapper::from(self.inner)
+                .sqrt_accurate()
+                .try_into_balance()?
+        } else {
+            self.inner.integer_sqrt()
+        };
+        Ok(Self::new(balance, self.is_divisible))
+    }
+}
+
+impl From<Balance> for BalanceUnit {
+    fn from(balance: Balance) -> Self {
+        Self {
+            inner: balance,
+            is_divisible: true,
+        }
+    }
+}
+
+impl Add for BalanceUnit {
+    type Output = Self;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        let (left, right) = match (self.is_divisible, rhs.is_divisible) {
+            (false, true) => (balance!(self.inner), rhs.inner),
+            (true, false) => (self.inner, balance!(rhs.inner)),
+            _ => (self.inner, rhs.inner),
+        };
+        let balance = left.saturating_add(right);
+        Self::new(balance, self.is_divisible || rhs.is_divisible)
+    }
+}
+
+impl AddAssign for BalanceUnit {
+    fn add_assign(&mut self, rhs: Self) {
+        *self = *self + rhs
+    }
+}
+
+impl Sub for BalanceUnit {
+    type Output = Self;
+
+    fn sub(self, rhs: Self) -> Self::Output {
+        let (left, right) = match (self.is_divisible, rhs.is_divisible) {
+            (false, true) => (balance!(self.inner), rhs.inner),
+            (true, false) => (self.inner, balance!(rhs.inner)),
+            _ => (self.inner, rhs.inner),
+        };
+        let balance = left.saturating_sub(right);
+        Self::new(balance, self.is_divisible || rhs.is_divisible)
+    }
+}
+
+impl SubAssign for BalanceUnit {
+    fn sub_assign(&mut self, rhs: Self) {
+        *self = *self - rhs
+    }
+}
+
+impl Mul for BalanceUnit {
+    type Output = Result<Self, ArithmeticError>;
+
+    fn mul(self, rhs: Self) -> Self::Output {
+        let balance = if self.is_divisible && rhs.is_divisible {
+            (FixedWrapper::from(self.inner) * (FixedWrapper::from(rhs.inner))).try_into_balance()?
```

Comment:
I don't like this idea, because:
- `mul` & `div` are special cases that could return an error, here we use `FixedWrapper`, but it is redundant for all other cases.
- `BalanceUnit` goes to storages as well. It is beeter to keep `u128` & `bool` than `Fixed` or `FixedWrapper`